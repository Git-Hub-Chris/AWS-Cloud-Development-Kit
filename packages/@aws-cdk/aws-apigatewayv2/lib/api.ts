import { Construct, IResource, Resource, Stack } from '@aws-cdk/core';

import { CfnApi } from './apigatewayv2.generated';
import { Deployment } from './deployment';
import { Integration } from './integration';
import { LambdaIntegration, LambdaIntegrationOptions } from './integrations/lambda-integration';
import { JsonSchema } from './json-schema';
import { Model, ModelOptions } from './model';
import { IRoute, KnownRouteSelectionExpression } from './route';
import { IStage, Stage, StageOptions } from './stage';

/**
 * Available protocols for ApiGateway V2 APIs (currently only 'WEBSOCKET' is supported)
 */
export enum ProtocolType {
  /**
   * WebSocket API
   */
  WEBSOCKET = "WEBSOCKET",

  /**
   * HTTP API
   */
  HTTP = "HTTP"
}

/**
 * Specifies how to interpret the base path of the API during import
 */
export enum BasePath {
  /**
   * Ignores the base path
   */
  IGNORE = "ignore",

  /**
   * Prepends the base path to the API path
   */
  PREPEND = "prepend",

  /**
   * Splits the base path from the API path
   */
  SPLIT = "split"
}

/**
 * This expression is evaluated when the service determines the given request should proceed
 * only if the client provides a valid API key.
 */
export enum KnownApiKeySelectionExpression {
  /**
   * Uses the `x-api-key` header to get the API Key
   */
  HEADER_X_API_KEY = "$request.header.x-api-key",

  /**
   * Uses the `usageIdentifier` property of the current authorizer to get the API Key
   */
  AUTHORIZER_USAGE_IDENTIFIER = " $context.authorizer.usageIdentifierKey"
}

/**
 * The BodyS3Location property specifies an S3 location from which to import an OpenAPI definition.
 */
export interface BodyS3Location {
  /**
   * The S3 bucket that contains the OpenAPI definition to import.
   */
  readonly bucket: string;

  /**
   * The Etag of the S3 object.
   *
   * @default - no etag verification
   */
  readonly etag?: string;

  /**
   * The key of the S3 object.
   */
  readonly key: string;

  /**
   * The version of the S3 object.
   *
   * @default - get latest version
   */
  readonly version?: string;
}

/**
 * The CorsConfiguration property specifies a CORS configuration for an API.
 */
export interface CorsConfiguration {
  /**
   * Specifies whether credentials are included in the CORS request.
   *
   * @default false
   */
  readonly allowCredentials?: boolean;

  /**
   * Represents a collection of allowed headers.
   *
   * @default - no headers allowed
   */
  readonly allowHeaders?: string[];

  /**
   * Represents a collection of allowed HTTP methods.
   *
   * @default - no allowed methods
   */
  readonly allowMethods: string[];

  /**
   * Represents a collection of allowed origins.
   *
   * @default - get allowed origins
   */
  readonly allowOrigins?: string[];

  /**
   * Represents a collection of exposed headers.
   *
   * @default - get allowed origins
   */
  readonly exposeHeaders?: string[];

  /**
   * The number of seconds that the browser should cache preflight request results.
   *
   * @default - get allowed origins
   */
  readonly maxAge?: number;
}

/**
 * Defines the contract for an Api Gateway V2 Api.
 */
export interface IApi extends IResource {
  /**
   * The ID of this API Gateway Api.
   * @attribute
   */
  readonly apiId: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Api.
 */
export interface ApiProps {
  /**
   * Indicates if a Deployment should be automatically created for this API,
   * and recreated when the API model (route, integration) changes.
   *
   * Since API Gateway deployments are immutable, When this option is enabled
   * (by default), an AWS::ApiGatewayV2::Deployment resource will automatically
   * created with a logical ID that hashes the API model (methods, resources
   * and options). This means that when the model changes, the logical ID of
   * this CloudFormation resource will change, and a new deployment will be
   * created.
   *
   * If this is set, `latestDeployment` will refer to the `Deployment` object
   * and `deploymentStage` will refer to a `Stage` that points to this
   * deployment. To customize the stage options, use the `deployOptions`
   * property.
   *
   * A CloudFormation Output will also be defined with the root URL endpoint
   * of this REST API.
   *
   * @default true
   */
  readonly deploy?: boolean;

  /**
   * Options for the API Gateway stage that will always point to the latest
   * deployment when `deploy` is enabled. If `deploy` is disabled,
   * this value cannot be set.
   *
   * @default - default options
   */
  readonly deployOptions?: StageOptions;

  /**
   * Retains old deployment resources when the API changes. This allows
   * manually reverting stages to point to old deployments via the AWS
   * Console.
   *
   * @default false
   */
  readonly retainDeployments?: boolean;

  /**
   * A name for the API Gateway Api resource.
   *
   * @default - ID of the Api construct.
   */
  readonly apiName?: string;

  /**
   * Specifies how to interpret the base path of the API during import.
   *
   * Supported only for HTTP APIs.
   * @default 'ignore'
   */
  readonly basePath?: BasePath;

  /**
   * The OpenAPI definition.
   *
   * To import an HTTP API, you must specify `body` or `bodyS3Location`.
   *
   * Supported only for HTTP APIs.
   * @default - `bodyS3Location` if defined, or no import
   */
  readonly body?: string;

  /**
   * The S3 location of an OpenAPI definition.
   *
   * To import an HTTP API, you must specify `body` or `bodyS3Location`.
   *
   * Supported only for HTTP APIs.
   * @default - `body` if defined, or no import
   */
  readonly bodyS3Location?: BodyS3Location;

  /**
   * The S3 location of an OpenAPI definition.
   *
   * To import an HTTP API, you must specify `body` or `bodyS3Location`.
   *
   * Supported only for HTTP APIs.
   * @default - `body` if defined, or no import
   */
  readonly corsConfiguration?: CorsConfiguration;

  /**
   * Available protocols for ApiGateway V2 APIs
   *
   * @default - required unless you specify an OpenAPI definition
   */
  readonly protocolType?: ProtocolType | string;

  /**
   * Expression used to select the route for this API
   *
   * @default - '${request.method} ${request.path}' for HTTP APIs, required for Web Socket APIs
   */
  readonly routeSelectionExpression?: KnownRouteSelectionExpression | string;

  /**
   * Expression used to select the Api Key to use for metering
   *
   * Supported only for WebSocket APIs
   *
   * @default - No Api Key
   */
  readonly apiKeySelectionExpression?: KnownApiKeySelectionExpression |Â string;

  /**
   * A description of the purpose of this API Gateway Api resource.
   *
   * @default - No description.
   */
  readonly description?: string;

  /**
   * Indicates whether schema validation will be disabled for this Api
   *
   * @default false
   */
  readonly disableSchemaValidation?: boolean;

  /**
   * Indicates the version number for this Api
   *
   * @default false
   */
  readonly version?: string;

  /**
   * Specifies whether to rollback the API creation (`true`) or not (`false`) when a warning is encountered
   *
   * @default false
   */
  readonly failOnWarnings?: boolean;
}

/**
 * Represents a WebSocket API in Amazon API Gateway v2.
 *
 * Use `addModel` and `addLambdaIntegration` to configure the API model.
 *
 * By default, the API will automatically be deployed and accessible from a
 * public endpoint.
 */
export class Api extends Resource implements IApi {

  /**
   * Creates a new imported API
   *
   * @param scope scope of this imported resource
   * @param id identifier of the resource
   * @param apiId Identifier of the API
   */
  public static fromApiId(scope: Construct, id: string, apiId: string): IApi {
    class Import extends Resource implements IApi {
      public readonly apiId = apiId;
    }

    return new Import(scope, id);
  }

  /**
   * The ID of this API Gateway Api.
   */
  public readonly apiId: string;

  /**
   * API Gateway stage that points to the latest deployment (if defined).
   */
  public deploymentStage?: Stage;

  protected resource: CfnApi;
  protected deployment?: Deployment;

  constructor(scope: Construct, id: string, props?: ApiProps) {
    if (props === undefined) {
      props = {};
    }
    super(scope, id, {
      physicalName: props.apiName || id,
    });

    if (props.protocolType === undefined && props.body === null && props.bodyS3Location === null) {
      throw new Error("You must specify a protocol type, or import an Open API definition (directly or from S3)");
    }

    switch (props.protocolType) {
      case ProtocolType.WEBSOCKET: {
        if (props.basePath !== undefined) {
          throw new Error('"basePath" is only supported with HTTP APIs');
        }
        if (props.body !== undefined) {
          throw new Error('"body" is only supported with HTTP APIs');
        }
        if (props.bodyS3Location !== undefined) {
          throw new Error('"bodyS3Location" is only supported with HTTP APIs');
        }
        if (props.corsConfiguration !== undefined) {
          throw new Error('"corsConfiguration" is only supported with HTTP APIs');
        }
        if (props.routeSelectionExpression === undefined) {
          throw new Error('"routeSelectionExpression" is required for Web Socket APIs');
        }
        break;
      }
      case ProtocolType.HTTP:
      case undefined: {
        if (props.apiKeySelectionExpression !== undefined) {
          throw new Error('"apiKeySelectionExpression" is only supported with Web Socket APIs');
        }
        if (props.disableSchemaValidation !== undefined) {
          throw new Error('"disableSchemaValidation" is only supported with Web Socket APIs');
        }
        if (props.routeSelectionExpression !== undefined && props.apiKeySelectionExpression !== KnownRouteSelectionExpression.METHOD_PATH) {
          throw new Error('"routeSelectionExpression" has a single supported value for HTTP APIs: "${request.method} ${request.path}"');
        }
        break;
      }
    }

    this.resource = new CfnApi(this, 'Resource', {
      ...props,
      name: this.physicalName
    });
    this.apiId = this.resource.ref;

    const deploy = props.deploy === undefined ? true : props.deploy;
    if (deploy) {
      // encode the stage name into the construct id, so if we change the stage name, it will recreate a new stage.
      // stage name is part of the endpoint, so that makes sense.
      const stageName = (props.deployOptions && props.deployOptions.stageName) || 'prod';

      this.deployment = new Deployment(this, 'Deployment', {
        api: this,
        description: 'Automatically created by the Api construct',

        // No stageName specified, this will be defined by the stage directly, as it will reference the deployment
        retainDeployments: props.retainDeployments
      });

      this.deploymentStage = new Stage(this, `Stage.${stageName}`, {
        ...props.deployOptions,
        deployment: this.deployment,
        api: this,
        stageName,
        description: 'Automatically created by the Api construct'
      });
    } else {
      if (props.deployOptions) {
        throw new Error(`Cannot set 'deployOptions' if 'deploy' is disabled`);
      }
    }
  }

  /**
   * API Gateway deployment that represents the latest changes of the API.
   * This resource will be automatically updated every time the REST API model changes.
   * This will be undefined if `deploy` is false.
   */
  public get latestDeployment() {
    return this.deployment;
  }

  /**
   * Defines an API Gateway Lambda integration.
   * @param id The construct id
   * @param props Lambda integration options
   */
  public addLambdaIntegration(id: string, props: LambdaIntegrationOptions): Integration {
    return new LambdaIntegration(this, id, { ...props, api: this });
  }

  /**
   * Defines a model for this Api Gateway.
   * @param schema The model schema
   * @param props The model integration options
   */
  public addModel(schema: JsonSchema, props?: ModelOptions): Model {
    return new Model(this, `Model.${schema.title}`, {
      ...props,
      modelName: schema.title,
      api: this,
      schema
    });
  }

  /**
   * Returns the ARN for a specific route and stage.
   *
   * @param route The route for this ARN ('*' if not defined)
   * @param stage The stage for this ARN (if not defined, defaults to the deployment stage if defined, or to '*')
   */
  public executeApiArn(route?: IRoute, stage?: IStage) {
    const stack = Stack.of(this);
    const apiId = this.apiId;
    const routeKey = ((route === undefined) ? '*' : route.key);
    const stageName = ((stage === undefined) ?
      ((this.deploymentStage === undefined) ? '*' : this.deploymentStage.stageName) :
      stage.stageName);
    return stack.formatArn({
      service: 'execute-api',
      resource: apiId,
      sep: '/',
      resourceName: `${stageName}/${routeKey}`
    });
  }

  /**
   * Returns the ARN for a specific connection.
   *
   * @param connectionId The identifier of this connection ('*' if not defined)
   * @param stage The stage for this ARN (if not defined, defaults to the deployment stage if defined, or to '*')
   */
  public connectionsApiArn(connectionId: string = "*", stage?: IStage) {
    const stack = Stack.of(this);
    const apiId = this.apiId;
    const stageName = ((stage === undefined) ?
      ((this.deploymentStage === undefined) ? '*' : this.deploymentStage.stageName) :
      stage.stageName);
    return stack.formatArn({
      service: 'execute-api',
      resource: apiId,
      sep: '/',
      resourceName: `${stageName}/POST/${connectionId}`
    });
  }

  /**
   * Returns the client URL for this Api, for a specific stage.
   *
   * Fails if `stage` is not defined, and `deploymentStage` is not set either by `deploy` or explicitly.
   * @param stage The stage for this URL (if not defined, defaults to the deployment stage)
   */
  public clientUrl(stage?: IStage): string {
    const stack = Stack.of(this);
    let stageName: string |Â undefined;
    if (stage === undefined) {
      if (this.deploymentStage === undefined) {
        throw Error("No stage defined for this Api");
      }
      stageName = this.deploymentStage.stageName;
    } else {
      stageName = stage.stageName;
    }
    return `wss://${this.apiId}.execute-api.${stack.region}.amazonaws.com/${stageName}`;
  }

  /**
   * Returns the connections URL for this Api, for a specific stage.
   *
   * Fails if `stage` is not defined, and `deploymentStage` is not set either by `deploy` or explicitly.
   * @param stage The stage for this URL (if not defined, defaults to the deployment stage)
   */
  public connectionsUrl(stage?: IStage): string {
    const stack = Stack.of(this);
    let stageName: string |Â undefined;
    if (stage === undefined) {
      if (this.deploymentStage === undefined) {
        throw Error("No stage defined for this Api");
      }
      stageName = this.deploymentStage.stageName;
    } else {
      stageName = stage.stageName;
    }
    return `https://${this.apiId}.execute-api.${stack.region}.amazonaws.com/${stageName}/@connections`;
  }
}