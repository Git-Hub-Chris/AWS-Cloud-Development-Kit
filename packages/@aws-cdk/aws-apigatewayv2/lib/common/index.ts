export * from './api-mapping';
export * from './authorizer';
export * from './domain-name';
export * from './integration';
export * from './json-schema';
export * from './route';
export * from './stage';