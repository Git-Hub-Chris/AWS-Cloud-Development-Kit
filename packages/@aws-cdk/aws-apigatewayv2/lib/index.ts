// AWS::APIGatewayv2 CloudFormation Resources:
export * from './apigatewayv2.generated';
export * from './api';
export * from './route';
export * from './integration';
export * from './integrations';
export * from './stage';