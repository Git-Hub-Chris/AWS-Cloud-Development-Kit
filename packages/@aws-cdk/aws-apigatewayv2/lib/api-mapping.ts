import { Construct, IResource, Resource } from '@aws-cdk/core';

import { IApi } from './api';
import { CfnApiMapping } from './apigatewayv2.generated';
import { IDomainName } from './domain-name';
import { IStage } from './stage';

/**
 * Defines the contract for an Api Gateway V2 Api Mapping.
 */
export interface IApiMapping extends IResource {
  /**
   * The ID of this API Gateway Api Mapping.
   * @attribute
   */
  readonly apiMappingId: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Api Mapping.
 *
 * This interface is used by the helper methods in `Api` and the sub-classes
 */
export interface ApiMappingOptions {
  /**
   * The API mapping key.
   *
   * @default - no API mapping key
   */
  readonly apiMappingKey?: string;

  /**
   * The associated domain name
   */
  readonly domainName: IDomainName |Â string;

  /**
   * The API stage.
   */
  readonly stage: IStage;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Api Mapping.
 */
export interface ApiMappingProps extends ApiMappingOptions {
  /**
   * Defines the api for this deployment.
   */
  readonly api: IApi;
}

/**
 * An Api Mapping for an API. An API mapping relates a path of your custom domain name to a stage of your API.
 *
 * A custom domain name can have multiple API mappings, but the paths can't overlap.
 *
 * A custom domain can map only to APIs of the same protocol type.
 */
export class ApiMapping extends Resource implements IApiMapping {

  /**
   * Creates a new imported API
   *
   * @param scope scope of this imported resource
   * @param id identifier of the resource
   * @param apiMappingId Identifier of the API Mapping
   */
  public static fromApiMappingId(scope: Construct, id: string, apiMappingId: string): IApiMapping {
    class Import extends Resource implements IApiMapping {
      public readonly apiMappingId = apiMappingId;
    }

    return new Import(scope, id);
  }

  /**
   * The ID of this API Gateway Api Mapping.
   */
  public readonly apiMappingId: string;

  protected resource: CfnApiMapping;

  constructor(scope: Construct, id: string, props: ApiMappingProps) {
    super(scope, id);

    this.resource = new CfnApiMapping(this, 'Resource', {
      apiId: props.api.apiId,
      domainName: ((typeof(props.domainName) === 'string') ? props.domainName : props.domainName.domainName),
      stage: props.stage.stageName,
      apiMappingKey: props.apiMappingKey,
    });
    this.apiMappingId = this.resource.ref;
  }
}