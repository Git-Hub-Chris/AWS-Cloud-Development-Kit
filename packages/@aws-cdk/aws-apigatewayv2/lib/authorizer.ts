import { Construct, IResource, Resource } from '@aws-cdk/core';

import { IApi } from './api';
import { CfnAuthorizer } from './apigatewayv2.generated';

/**
 * The authorizer type
 */
export enum AuthorizerType {
  /**
   * For WebSocket APIs, specify REQUEST for a Lambda function using incoming request parameters
   */
  REQUEST = 'REQUEST',

  /**
   * For HTTP APIs, specify JWT to use JSON Web Tokens
   */
  JWT = 'JWT'
}

/**
 * Specifies the configuration of a `JWT` authorizer.
 *
 * Required for the `JWT` authorizer type.
 *
 * Supported only for HTTP APIs.
 */
export interface JwtConfiguration {
  /**
   * A list of the intended recipients of the `JWT`.
   *
   * A valid `JWT` must provide an `aud` that matches at least one entry in this list.
   *
   * See RFC 7519.
   *
   * Supported only for HTTP APIs
   *
   * @default - no specified audience
   */
  readonly audience?: string[];

  /**
   * The base domain of the identity provider that issues JSON Web Tokens.
   *
   * For example, an Amazon Cognito user pool has the following format: `https://cognito-idp.{region}.amazonaws.com/{userPoolId}`.
   *
   * Required for the `JWT` authorizer type.
   *
   * Supported only for HTTP APIs.
   *
   * @default - no issuer
   */
  readonly issuer?: string;
}

/**
 * Defines the contract for an Api Gateway V2 Authorizer.
 */
export interface IAuthorizer extends IResource {
  /**
   * The ID of this API Gateway Api Mapping.
   * @attribute
   */
  readonly authorizerId: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Authorizer.
 *
 * This interface is used by the helper methods in `Api` and the sub-classes
 */
export interface AuthorizerOptions {
  /**
   * Specifies the required credentials as an IAM role for API Gateway to invoke the authorizer.
   *
   * To specify an IAM role for API Gateway to assume, use the role's Amazon Resource Name (ARN).
   *
   * To use resource-based permissions on the Lambda function, specify null. Supported only for `REQUEST` authorizers.
   *
   * @default - no credentials
   */
  readonly authorizerCredentialsArn?: string;

  /**
   * The authorizer type.
   */
  readonly authorizerType: AuthorizerType;

  /**
   * The authorizer's Uniform Resource Identifier (URI).
   *
   * For `REQUEST` authorizers, this must be a well-formed Lambda function URI, for example,
   * `arn:aws:apigateway:us-west-2:lambda:path/2015-03-31/functions/arn:aws:lambda:us-west-2:{acct_id}:function:{function_name}/invocations`.
   *
   * In general, the URI has this form: `arn:aws:apigateway:{region}:lambda:path/{service_api}`, where `{region}` is the same as the region
   * hosting the Lambda function, path indicates that the remaining substring in the URI should be treated as the path to the resource,
   * including the initial `/`. For Lambda functions, this is usually of the form `/2015-03-31/functions/{function_arn}/invocations`.
   */
  readonly authorizerUri: string;

  /**
   * The identity source for which authorization is requested.
   *
   * For a `REQUEST` authorizer, this is optional. The value is a set of one or more mapping expressions of the specified request parameters.
   * Currently, the identity source can be headers, query string parameters, stage variables, and context parameters.
   * For example, if an Auth header and a Name query string parameter are defined as identity sources, this value is `route.request.header.Auth`,
   * `route.request.querystring.Name`. These parameters will be used to perform runtime validation for Lambda-based authorizers by verifying all
   * of the identity-related request parameters are present in the request, not null, and non-empty. Only when this is true does the authorizer
   * invoke the authorizer Lambda function. Otherwise, it returns a 401 Unauthorized response without calling the Lambda function.
   *
   * For JWT, a single entry that specifies where to extract the JSON Web Token (JWT) from inbound requests. Currently only header-based and query
   * parameter-based selections are supported, for example `$request.header.Authorization`.
   *
   * @default - no identity source found
   */
  readonly identitySource?: string[];

  /**
   * The JWTConfiguration property specifies the configuration of a JWT authorizer.
   *
   * Required for the JWT authorizer type.
   *
   * Supported only for HTTP APIs.
   *
   * @default - only required for HTTP APIs
   */
  readonly jwtConfiguration?: JwtConfiguration;

  /**
   * The name of the authorizer.
   */
  readonly authorizerName: string;
}

/**
 * Defines the properties required for defining an Api Gateway V2 Authorizer.
 */
export interface AuthorizerProps extends AuthorizerOptions {
  /**
   * Defines the api for this deployment.
   */
  readonly api: IApi;
}

/**
 * An Api Mapping for an API. An API mapping relates a path of your custom domain name to a stage of your API.
 *
 * A custom domain name can have multiple API mappings, but the paths can't overlap.
 *
 * A custom domain can map only to APIs of the same protocol type.
 */
export class Authorizer extends Resource implements IAuthorizer {

  /**
   * Creates a new imported API
   *
   * @param scope scope of this imported resource
   * @param id identifier of the resource
   * @param authorizerId Identifier of the API Mapping
   */
  public static fromAuthorizerId(scope: Construct, id: string, authorizerId: string): IAuthorizer {
    class Import extends Resource implements IAuthorizer {
      public readonly authorizerId = authorizerId;
    }

    return new Import(scope, id);
  }

  /**
   * The ID of this API Gateway Api Mapping.
   */
  public readonly authorizerId: string;

  protected resource: CfnAuthorizer;

  constructor(scope: Construct, id: string, props: AuthorizerProps) {
    super(scope, id);

    this.resource = new CfnAuthorizer(this, 'Resource', {
      identitySource: (props.identitySource ? props.identitySource : []),
      apiId: props.api.apiId,
      name: props.authorizerName,
      authorizerType: props.authorizerType,
      authorizerCredentialsArn: props.authorizerCredentialsArn,
      // TODO: authorizerResultTtlInSeconds: props.authorizerResultTtl.toSeconds(),
      authorizerUri: props.authorizerUri,
      // TODO: identityValidationExpression: props.identityValidationExpression
      jwtConfiguration: props.jwtConfiguration,
    });
    this.authorizerId = this.resource.ref;
  }
}