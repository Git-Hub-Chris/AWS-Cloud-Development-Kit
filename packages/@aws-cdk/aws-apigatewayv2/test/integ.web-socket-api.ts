import * as lambda from '@aws-cdk/aws-lambda';
import * as cdk from '@aws-cdk/core';
import * as apigatewayv2 from '../lib';

// tslint:disable:max-line-length

const app = new cdk.App();

const stack = new cdk.Stack(app, 'ApiagtewayV2HttpApi');

const webSocketHandler = new lambda.Function(stack, 'MyFunc', {
  runtime: lambda.Runtime.PYTHON_3_7,
  handler: 'index.handler',
  code: new lambda.InlineCode(`
import json
def handler(event, context):
      return {
        'status': 'ok',
        'message': 'success'
      }`),
});

// create a Web Socket API
const api = new apigatewayv2.WebSocketApi(stack, 'WebSocketApi', {
  routeSelectionExpression: '${request.body.action}',
  deployOptions: {
    stageName: 'integration',
  },
});

const defaultStatusIntegrationResponse: apigatewayv2.WebSocketIntegrationResponseOptions = {
  responseTemplates: {
    default: '#set($inputRoot = $input.path(\'$\')) { "status": "${inputRoot.status}", "message": "$util.escapeJavaScript(${inputRoot.message})" }',
  },
  templateSelectionExpression: 'default',
};

const defaultStatusRouteResponse: apigatewayv2.WebSocketRouteResponseOptions = {
  modelSelectionExpression: 'default',
  responseModels: {
    default: api.addModel({ schema: apigatewayv2.JsonSchemaVersion.DRAFT4, title: 'statusResponse', type: apigatewayv2.JsonSchemaType.OBJECT, properties: { status: { type: apigatewayv2.JsonSchemaType.STRING }, message: { type: apigatewayv2.JsonSchemaType.STRING } } }),
  },
};

const webSocketConnectIntegration = api.addLambdaIntegration('default', {
  handler: webSocketHandler,
  proxy: false,
  passthroughBehavior: apigatewayv2.WebSocketPassthroughBehavior.NEVER,
  requestTemplates: {
    connect: '{ "action": "${context.routeKey}", "userId": "${context.identity.cognitoIdentityId}", "connectionId": "${context.connectionId}", "domainName": "${context.domainName}", "stageName": "${context.stage}" }',
  },
  templateSelectionExpression: 'connect',
  description: 'WebSocket Api Connection Integration',
});
webSocketConnectIntegration.addResponse(apigatewayv2.WebSocketKnownRouteResponseKey.DEFAULT, defaultStatusIntegrationResponse);

api.addRoute(apigatewayv2.WebSocketKnownRouteKey.CONNECT, webSocketConnectIntegration, {
  authorizationType: apigatewayv2.WebSocketAuthorizationType.IAM,
  routeResponseSelectionExpression: apigatewayv2.WebSocketKnownRouteResponseKey.DEFAULT,
}).addResponse(apigatewayv2.WebSocketKnownRouteResponseKey.DEFAULT, defaultStatusRouteResponse);
