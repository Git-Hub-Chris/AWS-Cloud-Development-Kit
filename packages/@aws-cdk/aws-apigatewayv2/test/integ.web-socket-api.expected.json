{
  "Resources": {
    "MyFuncServiceRole54065130": {
      "Type": "AWS::IAM::Role",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com"
              }
            }
          ],
          "Version": "2012-10-17"
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition"
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
              ]
            ]
          }
        ]
      }
    },
    "MyFunc8A243A2C": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
        "Code": {
          "ZipFile": "\nimport json\ndef handler(event, context):\n      return {\n        'status': 'ok',\n        'message': 'success'\n      }"
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "MyFuncServiceRole54065130",
            "Arn"
          ]
        },
        "Runtime": "python3.7"
      },
      "DependsOn": [
        "MyFuncServiceRole54065130"
      ]
    },
    "MyFuncApiPermissionApiagtewayV2HttpApiWebSocketApidefaultlambdaintegration33FE346D3DE42898": {
      "Type": "AWS::Lambda::Permission",
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "MyFunc8A243A2C",
            "Arn"
          ]
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region"
              },
              ":",
              {
                "Ref": "AWS::AccountId"
              },
              ":",
              {
                "Ref": "WebSocketApi34BCF99B"
              },
              "/",
              {
                "Ref": "WebSocketApiDefaultStage734E750B"
              },
              "/$connect"
            ]
          ]
        }
      }
    },
    "WebSocketApi34BCF99B": {
      "Type": "AWS::ApiGatewayV2::Api",
      "Properties": {
        "Name": "WebSocketApi",
        "ProtocolType": "WEBSOCKET",
        "RouteSelectionExpression": "${request.body.action}"
      }
    },
    "WebSocketApiDeployment507BF677": {
      "Type": "AWS::ApiGatewayV2::Deployment",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "Description": "Automatically created by the Api construct"
      },
      "DependsOn": [
        "WebSocketApiconnectrouteA20DAA0E",
        "WebSocketApidefaultlambdaintegration33A4E43E",
        "WebSocketApiModelstatusResponseB7B1089E"
      ]
    },
    "WebSocketApiDefaultStage734E750B": {
      "Type": "AWS::ApiGatewayV2::Stage",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "StageName": "integration",
        "DeploymentId": {
          "Ref": "WebSocketApiDeployment507BF677"
        },
        "Description": "Automatically created by the Api construct"
      }
    },
    "WebSocketApiModelstatusResponseB7B1089E": {
      "Type": "AWS::ApiGatewayV2::Model",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "Name": "statusResponse",
        "Schema": {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "title": "statusResponse",
          "type": "object",
          "properties": {
            "status": {
              "type": "string"
            },
            "message": {
              "type": "string"
            }
          }
        },
        "ContentType": "application/json"
      }
    },
    "WebSocketApidefaultlambdaintegration33A4E43E": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "IntegrationType": "AWS",
        "Description": "WebSocket Api Connection Integration",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition"
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region"
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "MyFunc8A243A2C",
                  "Arn"
                ]
              },
              "/invocations"
            ]
          ]
        },
        "PassthroughBehavior": "NEVER",
        "RequestTemplates": {
          "connect": "{ \"action\": \"${context.routeKey}\", \"userId\": \"${context.identity.cognitoIdentityId}\", \"connectionId\": \"${context.connectionId}\", \"domainName\": \"${context.domainName}\", \"stageName\": \"${context.stage}\" }"
        },
        "TemplateSelectionExpression": "connect"
      }
    },
    "WebSocketApidefaultlambdaintegrationResponsedefault8A571455": {
      "Type": "AWS::ApiGatewayV2::IntegrationResponse",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "IntegrationId": {
          "Ref": "WebSocketApidefaultlambdaintegration33A4E43E"
        },
        "IntegrationResponseKey": "$default",
        "ResponseTemplates": {
          "default": "#set($inputRoot = $input.path('$')) { \"status\": \"${inputRoot.status}\", \"message\": \"$util.escapeJavaScript(${inputRoot.message})\" }"
        },
        "TemplateSelectionExpression": "default"
      }
    },
    "WebSocketApiconnectrouteA20DAA0E": {
      "Type": "AWS::ApiGatewayV2::Route",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "RouteKey": "$connect",
        "AuthorizationType": "AWS_IAM",
        "RouteResponseSelectionExpression": "$default",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "WebSocketApidefaultlambdaintegration33A4E43E"
              }
            ]
          ]
        }
      }
    },
    "WebSocketApiconnectrouteResponsedefaultB361BECB": {
      "Type": "AWS::ApiGatewayV2::RouteResponse",
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B"
        },
        "RouteId": {
          "Ref": "WebSocketApiconnectrouteA20DAA0E"
        },
        "RouteResponseKey": "$default",
        "ModelSelectionExpression": "default",
        "ResponseModels": {
          "default": "statusResponse"
        }
      },
      "DependsOn": [
        "WebSocketApiModelstatusResponseB7B1089E"
      ]
    }
  }
}